### PR TITLE
bpo-30594: Add a test for bad IDNA in ssl server_hostname

### PR DESCRIPTION
**Pull request by njsmith**
See discussion:
  https://github.com/python/cpython/pull/1992#issuecomment-307024778

### CODE REVIEW

**Comment by njsmith** [comment]
CC @haypo @serhiy-storchaka

**Review by serhiy-storchaka — APPROVED** [review]
LGTM. Thank you for the test.

File: Lib/test/test_ssl.py [hunk 1]
```diff
@@ -1393,6 +1393,18 @@ def test_subclass(self):
...
 
+    def test_bad_idna_in_server_hostname(self):
+        # Note: this test is testing some code that probably shouldn't exist
+        # in the first place, so if it starts failing at some point because
+        # you made the ssl module stop doing IDNA decoding then please feel
+        # free to remove it.
+        ctx = ssl.create_default_context()
+        # Do it a few times, because this used to cause memory corruption
+        # https://bugs.python.org/issue30594
+        for _ in range(5):
```
Comment (vstinner) [inline Lib/test/test_ssl.py]: I don't think that it's worth it to run it in a loop. Just run it once. It's only a regression test. We have enough buildbots to catch race conditions *if* there is a test ;-)

File: Lib/test/test_ssl.py [hunk 2]
```diff
@@ -1393,6 +1393,18 @@ def test_subclass(self):
...
                 self.assertEqual(cm.exception.errno, ssl.SSL_ERROR_WANT_READ)
 
+    def test_bad_idna_in_server_hostname(self):
+        # Note: this test is testing some code that probably shouldn't exist
+        # in the first place, so if it starts failing at some point because
+        # you made the ssl module stop doing IDNA decoding then please feel
+        # free to remove it.
+        ctx = ssl.create_default_context()
+        # Do it a few times, because this used to cause memory corruption
+        # https://bugs.python.org/issue30594
```
Comment (vstinner) [inline Lib/test/test_ssl.py]: Usually, we just write "# bpo-30594: ..."

**Comment by njsmith** [comment]
Comments addressed.

**Comment by vstinner** [comment]
Thanks @njsmith for the unit test! I merged it.

**Pull request merged**